src/Microsoft.AspNetCore.Razor.Tools/MetadataCache.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Tools
+{
+    internal class MetadataCache
+    {
+        // Store 500 entries -- arbitrary number
+        private const int CacheSize = 500;
+        private readonly ConcurrentLruCache<string, MetadataCacheEntry> _metadataCache =
+            new ConcurrentLruCache<string, MetadataCacheEntry>(CacheSize, StringComparer.OrdinalIgnoreCase);
+
+        // For testing purposes only.
+        internal ConcurrentLruCache<string, MetadataCacheEntry> Cache => _metadataCache;
+
+        internal Metadata GetMetadata(string fullPath)
+        {
+            var timestamp = GetFileTimeStamp(fullPath);
+
+            // Check if we have an entry in the dictionary.
+            if (_metadataCache.TryGetValue(fullPath, out var entry))
+            {
+                if (timestamp.HasValue && timestamp.Value == entry.Timestamp)
+                {
+                    return entry.Metadata;
+                }
+                else
+                {
+                    _metadataCache.Remove(fullPath);
+                }
+            }
+
+            using (var fileStream = File.OpenRead(fullPath))
+            {
+                var metadata = AssemblyMetadata.CreateFromStream(fileStream, PEStreamOptions.PrefetchMetadata);
+
+                var result = _metadataCache.GetOrAdd(fullPath, new MetadataCacheEntry(timestamp.Value, metadata));
```

Comment:
Let's move this outside the `using` block. `GetOrAdd` has a very coarse lock so you might have end up keeping the open for long periods

Reply:
I think we should just the created `metadata` here and ignore `result`. There's a possible race here where `result` could be older than `metadata`(T1 -> Calls remove -> Makes metadata -> Does GetOAdd. Meanwhile T2 -> Fails TryGetValue and lands here. GetOrAddr returns results from T1 which might be older). The cache entry should get updated in the next call to this method, so we don't super need to worry about fixing that.